AI/BattleAI/AttackPossibility.cpp

```diff
@@ -9,89 +9,184 @@
+		if(defHex == hex) // should be impossible but check anyway
+			continue;
+
+		AttackPossibility ap(hex, defHex, attackInfo);
+		ap.attackerState = attacker->acquireState();
+		ap.shootersBlockedDmg = bestAp.shootersBlockedDmg;
 
-		vstd::amin(attackDmg.first, defenderState->getAvailableHealth());
-		vstd::amin(attackDmg.second, defenderState->getAvailableHealth());
+		const int totalAttacks = ap.attackerState->getTotalAttacks(attackInfo.shooting);
 
-		vstd::amin(retaliation.first, ap.attackerState->getAvailableHealth());
-		vstd::amin(retaliation.second, ap.attackerState->getAvailableHealth());
+		if (!attackInfo.shooting)
+			ap.attackerState->setPosition(hex);
 
-		ap.damageDealt += (attackDmg.first + attackDmg.second) / 2;
+		std::vector<const battle::Unit*> units;
 
-		ap.attackerState->afterAttack(attackInfo.shooting, false);
+		if (attackInfo.shooting)
+			units = state->getAttackedBattleUnits(attacker, defHex, true, BattleHex::INVALID);
+		else
+			units = state->getAttackedBattleUnits(attacker, defHex, false, hex);
 
-		//FIXME: use ranged retaliation
-		if(!attackInfo.shooting && defenderState->ableToRetaliate() && !counterAttacksBlocked)
+		// ensure the defender is also affected
+		bool addDefender = true;
+		for(auto unit : units)
 		{
-			ap.damageReceived += (retaliation.first + retaliation.second) / 2;
-			defenderState->afterAttack(attackInfo.shooting, true);
+			if (unit->unitId() == defender->unitId())
+			{
+				addDefender = false;
+				break;
+			}
 		}
 
-		ap.attackerState->damage(ap.damageReceived);
-		defenderState->damage(ap.damageDealt);
+		if(addDefender)
+			units.push_back(defender);
 
-		if(!ap.attackerState->alive() || !defenderState->alive())
-			break;
+		for(auto u : units)
+		{
+			if(!ap.attackerState->alive())
+				break;
+
+			assert(u->unitId() != attacker->unitId());
+
+			auto defenderState = u->acquireState();
+			ap.affectedUnits.push_back(defenderState);
+
+			for(int i = 0; i < totalAttacks; i++) {
+				si64 damageDealt, damageReceived;
+
+				TDmgRange retaliation(0, 0);
+				auto attackDmg = getCbc()->battleEstimateDamage(ap.attack, &retaliation);
+
+				vstd::amin(attackDmg.first, defenderState->getAvailableHealth());
+				vstd::amin(attackDmg.second, defenderState->getAvailableHealth());
+
+				vstd::amin(retaliation.first, ap.attackerState->getAvailableHealth());
+				vstd::amin(retaliation.second, ap.attackerState->getAvailableHealth());
+
+				damageDealt = (attackDmg.first + attackDmg.second) / 2;
+				ap.attackerState->afterAttack(attackInfo.shooting, false);
+
+				//FIXME: use ranged retaliation
+				damageReceived = 0;
+				if (!attackInfo.shooting && defenderState->ableToRetaliate() && !counterAttacksBlocked)
+				{
+					damageReceived = (retaliation.first + retaliation.second) / 2;
+					defenderState->afterAttack(attackInfo.shooting, true);
+				}
+
+				bool isEnemy = state->battleMatchOwner(attacker, u) && !mindControlled;
```

Comment:
Whole point of `battleMatchOwner` is to take mind control in account. If you ignore it, just compare sides.
```suggestion
				bool isEnemy = attacker->unitSide() != u->unitSide();
```
And `mindControlled` calculation could be removed too. 
BTW. Any bonus system usage in battle evaluation is known performance bottleneck.

Reply:
I guess this part could be refactored into something as follows:

```
bool isEnemy = isRealEnemy(u); // meaning it's the opposite side to the player controlling the attacking unit
if (isEnemy)
ap.damageDealt += damage;
if (!isEnemy || mindControlled(u))
ap.collateralDamage += damage;
```

Any help would be appreciated.

Reply:
We are evaluating from PoV of player controlling attacker unit, so we compare side of effective owner to raw side of defender. Smth like:
```cpp
const auto attackerSide = getCbc()->playerToSide(getCbc()->battleGetOwner(attacker));//from your lambda, replaces it
//...
const bool isEnemy = attackerSide != u->unitSide();
if(isEnemy)
ap.damageDealt += damage;
else
ap.collateralDamage += damage;
```